libpromises/evalfunction.c

```diff
@@ -1674,6 +1680,211 @@ static FnCallResult FnCallSplayClass(EvalContext *ctx,
+                               const Rlist *finalargs)
+{
+#ifdef HAVE_LIBCURL
+    char *url = RlistScalarValue(finalargs);
+    VarRef *ref = ResolveAndQualifyVarName(fp, RlistScalarValue(finalargs->next));
+    if (!ref)
+    {
+        return FnFailure();
+    }
+
+    JsonElement *options = VarRefValueToJson(ctx, fp, ref, NULL, 0);
+    VarRefDestroy(ref);
+
+    if (JsonGetElementType(options) != JSON_ELEMENT_TYPE_CONTAINER ||
+        JsonGetContainerType(options) != JSON_CONTAINER_TYPE_OBJECT)
+    {
+        return FnFailure();
+    }
+
+    Writer *cache_w = StringWriter();
+    WriterWriteF(cache_w, "url = %s; options = ", url);
+    JsonWriteCompact(cache_w, options);
+
+    if (NULL == CURL_CACHE)
+    {
+        CURL_CACHE = JsonObjectCreate(10);
+    }
+    atexit(&CurlCleanup);
+
+    JsonElement *old_result = JsonObjectGetAsObject(CURL_CACHE, StringWriterData(cache_w));
+
+    if (NULL != old_result)
+    {
+        Log(LOG_LEVEL_VERBOSE, "%s: found cached request for %s", fp->name, url);
+        WriterClose(cache_w);
+        return (FnCallResult) { FNCALL_SUCCESS, (Rval) { JsonCopy(old_result), RVAL_TYPE_CONTAINER } };
+    }
+
+    if (!CURL_INITIALIZED && curl_global_init(CURL_GLOBAL_DEFAULT) != 0)
+    {
+        Log(LOG_LEVEL_ERR, "%s: libcurl initialization failed, sorry", fp->name);
+
+        WriterClose(cache_w);
+        return FnFailure();
+    }
+
+    CURL_INITIALIZED = true;
+
+    CURL *curl = curl_easy_init();
+    if (!curl)
+    {
+        Log(LOG_LEVEL_ERR, "%s: libcurl easy_init failed, sorry", fp->name);
+
+        WriterClose(cache_w);
+        return FnFailure();
+    }
+
+    Buffer *content = BufferNew();
+    Buffer *headers = BufferNew();
+    curl_easy_setopt(curl, CURLOPT_URL, url);
+    curl_easy_setopt(curl, CURLOPT_NOSIGNAL, 1); // do not use signals
+    curl_easy_setopt(curl, CURLOPT_TIMEOUT, 3L); // set default timeout
+    curl_easy_setopt(curl, CURLOPT_VERBOSE, 0);
+    curl_easy_setopt(curl,
+                     CURLOPT_PROTOCOLS,
+
+                     // Allowed protocols
+                     CURLPROTO_FILE |
+                     CURLPROTO_FTP |
+                     CURLPROTO_FTPS |
+                     CURLPROTO_HTTP |
+                     CURLPROTO_HTTPS);
+
+    curl_easy_setopt(curl, CURLOPT_WRITEFUNCTION, cfengine_curl_write_callback);
+    curl_easy_setopt(curl, CURLOPT_WRITEDATA, content);
+    curl_easy_setopt(curl, CURLOPT_HEADERFUNCTION, cfengine_curl_write_callback);
+    curl_easy_setopt(curl, CURLOPT_HEADERDATA, headers);
+
+    size_t max_content = 4096;
+    size_t max_headers = 4096;
+    JsonIterator iter = JsonIteratorInit(options);
+    const JsonElement *e;
+    while ((e = JsonIteratorNextValueByType(&iter, JSON_ELEMENT_TYPE_PRIMITIVE, true)))
+    {
+        const char *key = JsonIteratorCurrentKey(&iter);
+        const char *value = JsonPrimitiveGetAsString(e);
+
+        if (0 == strcmp(key, "curl.timeout"))
```

Comment:
Like the function name, I don't see anything curl specific about these options, so a name like ``http.timeout`` might be better. OTOH, I see that other protocols are allowed too, so maybe something like ``web_get`` and ``web.timeout``? I'm open to suggestions. :-)